src/pykeen/nn/utils.py

```diff
@@ -185,3 +190,93 @@ def safe_diagonal(matrix: torch.Tensor) -> torch.Tensor:
+
+
+StackDirection = Literal["horizontal", "vertical"]
+HORIZONTAL: StackDirection = "horizontal"
+VERTICAL: StackDirection = "vertical"
+
+
+def decide_stacking_direction(
+    input_dim: int,
+    output_dim: int,
+) -> StackDirection:
+    """
+    Determine a stacking direction based on the input and output dimension.
+
+    The vertical stacking approach is suitable for low dimensional input and high dimensional output,
+    because the projection to low dimensions is done first. While the horizontal stacking approach is good
+    for high dimensional input and low dimensional output as the projection to high dimension is done last.
+
+    :param input_dim:
+        the layer's input dimension
+    :param output_dim:
+        the layer's output dimension
+
+    :return:
+        the suggested stacking direction
+
+    .. seealso :: [thanapalasingam2021]_
+    """
+    if input_dim > output_dim:
+        return HORIZONTAL
+    return VERTICAL
+
+
+def adjacency_tensor_to_stacked_matrix(
+    num_relations: int,
+    num_entities: int,
+    source: torch.LongTensor,
+    target: torch.LongTensor,
+    edge_type: torch.LongTensor,
+    edge_weights: Optional[torch.FloatTensor] = None,
+    direction: StackDirection = HORIZONTAL,
+) -> torch.Tensor:
+    """
+    Stack adjacency matrices as described in [thanapalasingam2021]_.
+
+    This method re-arranges the (sparse) adjacency tensor of shape
+    `(num_entities, num_relations, num_entities)` to a sparse adjacency matrix of shape
+    `(num_entities, num_relations * num_entities)` (horizontal stacking) or
+    `(num_entities * num_relations, num_entities)` (vertical stacking). Thereby, we can perform the relation-specific
+    message passing of R-GCN by a single sparse matrix multiplication (and some additional pre- and/or
+    post-processing) of the inputs.
+
+    :param num_relations:
+        the number of relations
+    :param num_entities:
+        the number of entities
+    :param source: shape: (num_triples,)
+        the source entity indices
+    :param target: shape: (num_triples,)
+        the target entity indices
+    :param edge_type: shape: (num_triples,)
+        the edge type, i.e., relation ID
+    :param edge_weights: shape: (num_triples,)
+        scalar edge weights
+    :param direction:
+        whether to use horizontal or vertical stacking
+
+    :return: shape: `(num_entities * num_relations, num_entities)` or `(num_entities, num_entities * num_relations)`
+        the stacked adjacency matrix
+
+    :raises ValueError:
+        if the direction is invalid
+    """
+    offset = edge_type * num_entities
+    if direction == HORIZONTAL:
+        size = (num_entities, num_relations * num_entities)
+        target = offset + target
+    elif direction == VERTICAL:
+        size = (num_relations * num_entities, num_entities)
+        source = offset + source
+    else:
+        raise ValueError(f"Invalid stacking direction: {direction}")
```

Comment:
is it possible to have a 3rd in the future?

Reply:
No, I don't think so: since the adjacency matrix is 2d, there are only two meaningful stacking axes.

Reply:
then wouldn't it be simpler to just use a bool like before?

Reply:
7b3e321